pkg/typesystem/resolver.go

```diff
@@ -0,0 +1,79 @@
+package typesystem
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"time"
+
+	"github.com/karlseguin/ccache/v3"
+	"github.com/oklog/ulid/v2"
+	"github.com/openfga/openfga/pkg/storage"
+)
+
+const (
+	typesystemCacheTTL = 168 * time.Hour // 7 days
+)
+
+// TypesystemResolverFunc is a function that implementations can implement to provide lookup and
+// resolution of a Typesystem.
+type TypesystemResolverFunc func(ctx context.Context, storeID, modelID string) (*TypeSystem, error)
+
+// MemoizedTypesystemResolverFunc returns a TypesystemResolverFunc that either fetches the provided authorization
+// model (if provided) or looks up the latest authorizaiton model, and then it constructs a TypeSystem from
+// the resolved model. The type-system resolution is memoized so if another lookup of the same model occurs,
+// then the earlier TypeSystem that was constructed will be used.
+//
+// The memoized resolver function is safe for concurrent use.
+func MemoizedTypesystemResolverFunc(datastore storage.AuthorizationModelReadBackend) TypesystemResolverFunc {
+
+	cache := ccache.New(ccache.Configure[*TypeSystem]())
```

Comment:
also, we need to call `cache.Stop()` when the server terminates - this is important when running the tests

Reply:
You have to call `.Stop()` on an in-memory cache. That's funky as hell and not cool.. What's the issue in the tests, is it the goroutine leak because of the background cache worker?

Reply:
https://github.com/karlseguin/ccache/issues/59 seems that if you call stop you get races potentially as well..

Reply:
Maybe we should look for an alternative cache package? The cache design is pretty odd to begin with TBH.

https://github.com/allegro/bigcache is widely used for high concurrency scenarios, which we have in this case.

Reply:
> is it the goroutine leak because of the background cache worker? Is that just because of our usage of uber-go/goleak?

i haven't ran a test with it recently but i expect to see some leaks, yeah. Not a big deal though because it would only affect tests

Reply:
We'd have to change composition a good amount so we can introduce a `.Close` or `.Stop` method on a struct that implements some `TypesystemResolver` interface or order to manage this. It feels like a really dirty pattern just so we can call close/stop on the underlying cache. An in-memory cache really shouldn't have a Stop...

I'd vote we table it and look at swapping away from our existing cache lib, it's odd.